packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,843 @@
+          ? Effect<E, L, R>.from(
+              (context) => _unsafeRun(context).then(
+                (exit) => switch (exit) {
+                  Left(value: final value) => Left(value),
+                  Right(value: final value) =>
+                    env.closeScope<E, L>()._unsafeRun(context).then(
+                          (exit) => switch (exit) {
+                            Left(value: final value) => Left(value),
+                            Right() => Right(value),
+                          },
+                        ),
+                },
+              ),
+            )
+          : this;
+
+  /// {@category context}
+  Effect<Null, L, R> provide(Context<E> context) => Effect.from(
+        (_) => _provideEnvCloseScope(context.env)._unsafeRun(context),
+      );
+
+  /// {@category context}
+  Effect<Null, L, R> provideEnv(E env) => Effect.from(
+        (_) => _provideEnvCloseScope(env)._unsafeRun(Context.env(env)),
+      );
+
+  /// {@category context}
+  Effect<V, L, R> provideEffect<V>(Effect<V, L, E> effect) => Effect.from(
+        (context) => effect._unsafeRun(context).then(
+              (exit) => switch (exit) {
+                Left(value: final cause) => Left(cause),
+                Right(value: final value) => _unsafeRun(context.withEnv(value)),
+              },
+            ),
+      );
+
+  /// {@category context}
+  static Effect<E, L, E> env<E, L>() => Effect.from(
+        (context) => Right(context.env),
+      );
+
+  /// {@category combining}
+  Effect<E, L, V> ap<V>(
+    Effect<E, L, V Function(R r)> f,
+  ) =>
+      f.flatMap(
+        (f) => flatMap(
+          (v) => Effect.succeed(f(v)),
+        ),
+      );
+
+  /// {@category conversions}
+  Effect<E, Never, Either<L, R>> either() => Effect.from(
+        (context) => _unsafeRun(context).then(
+          (exit) => switch (exit) {
+            Left(value: final cause) => switch (cause) {
+                Failure<L>(error: final error) => Right(Left(error)),
+                Die() => Left(cause),
+                Interrupted() => Left(cause),
+              },
+            Right(value: final value) => Right(Right(value)),
+          },
+        ),
+      );
+
+  /// {@category conversions}
+  Effect<E, Never, Option<R>> option() => Effect.from(
+        (context) => _unsafeRun(context).then(
+          (exit) => switch (exit) {
+            Left(value: final cause) => switch (cause) {
+                Failure() => Right<Cause<Never>, Option<R>>(const None()),
+                Die() => Left(cause),
+                Interrupted() => Left(cause),
+              },
+            Right(value: final value) => Right(Some(value)),
+          },
+        ),
+      );
+
+  /// {@category conversions}
+  Effect<E, Never, Exit<L, R>> exit() => Effect.from(
+        (context) => _unsafeRun(context).then(
+          (exit) => Right(exit),
+        ),
+      );
+
+  /// {@category folding}
+  Effect<E, Never, C> match<C>({
```

Comment:
This has a semi-non-trivial definition, so I'm hesitant to suggest this, but maybe it'd be possible to use native pattern matching? i.e.

```dart
final val = switch ($.syncEither(effect)) {
  Left() => ...,
  Right() => ...,
}
```

rather than

```dart
final val = effect.match(
  () => ...,
  () => ...,
)
```

Yes, it's more verbose, but the native pattern matching feels much more "right" to me. The function feels like it's working around an issue from JS that isn't in Dart.